src/loops.rs

```diff
@@ -92,6 +97,35 @@ impl LintPass for LoopsPass {
                     }
                 }
             }
+
+            // Look for variables that are incremented once per loop iteration.
+            let mut visitor = IncrementVisitor { cx: cx, states: HashMap::new(), depth: 0, done: false };
+            walk_expr(&mut visitor, body);
+
+            // For each candidate, check the parent block to see if
+            // it's initialized to zero at the start of the loop.
+            let map = &cx.tcx.map;
+            let parent_scope = map.get_enclosing_scope(expr.id).and_then(|id| map.get_enclosing_scope(id) );
+            if let Some(parent_id) = parent_scope {
+                if let NodeBlock(block) = map.get(parent_id) {
+                    for (id, _) in visitor.states.iter().filter( |&(_,v)| *v == VarState::IncrOnce) {
+                        let mut visitor2 = InitializeVisitor { cx: cx, end_expr: expr, var_id: id.clone(),
+                                                               state: VarState::IncrOnce, name: None,
+                                                               depth: 0, done: false };
+                        walk_block(&mut visitor2, block);
+
+                        if visitor2.state == VarState::Warn {
+                            if let Some(name) = visitor2.name {
+                                span_lint(cx, EXPLICIT_COUNTER_LOOP, expr.span,
+                                          &format!("the variable `{0}` is used as a loop counter. Consider \
+                                                    using `for ({0}, item) in _.iter().enumerate()` \
```

Comment:
Right, I was meaning to ask about that. How do you get from an Expr to the corresponding source? I couldn't find an example of this -- a lot of other lints just include variable names.